src/Symfony/Component/Messenger/DependencyInjection/MessengerPass.php

```diff
@@ -102,27 +102,25 @@ private function registerHandlers(ContainerBuilder $container)
         $handlerResolver->replaceArgument(0, ServiceLocatorTagPass::register($container, $handlersLocatorMapping));
     }
 
-    private function guessHandledClass(ContainerBuilder $container, string $serviceId): string
+    private function guessHandledClass(\ReflectionClass $handlerClass, string $serviceId): string
     {
-        $reflection = $container->getReflectionClass($container->getDefinition($serviceId)->getClass());
-
         try {
-            $method = $reflection->getMethod('__invoke');
+            $method = $handlerClass->getMethod('__invoke');
         } catch (\ReflectionException $e) {
-            throw new RuntimeException(sprintf('Service "%s" should have an `__invoke` function.', $serviceId));
+            throw new RuntimeException(sprintf('Invalid handler service "%s": class "%s" must have an "__invoke()" method.', $serviceId, $handlerClass->getName()));
         }
 
         $parameters = $method->getParameters();
         if (1 !== count($parameters)) {
-            throw new RuntimeException(sprintf('`__invoke` function of service "%s" must have exactly one parameter.', $serviceId));
+            throw new RuntimeException(sprintf('Invalid handler service "%s": method "%s::__invoke()" must have exactly one argument corresponding to the message it handles.', $serviceId, $handlerClass->getName())
+        );
         }
 
-        $parameter = $parameters[0];
-        if (null === $parameter->getClass()) {
-            throw new RuntimeException(sprintf('The parameter of `__invoke` function of service "%s" must type hint the message class it handles.', $serviceId));
+        if (!$parameters[0]->hasType()) {
```

Comment:
should also report the issue if it has a builtin type as typehint (i.e. not a class)